extensions/amp-auto-ads/0.1/placement.js

```diff
@@ -196,6 +212,45 @@ export class Placement {
     });
   }
 
+  /**
+   * Gets instructions for the placement in terms of height, width and margins.
+   * If responsive is on, ad should be placed at full viewport width and a
+   * proportionate height, and the margins are adjusted so that the ad edges
+   * stick to both ends of the viewport.
+   * @param {!./ad-network-config.SizeInfoDef} sizing
+   * @return {PlacementDef}
+   * @private
+   */
+  getPlacement_(sizing) {
+    if (isExperimentOn(this.ampdoc.win, 'amp-auto-ads-responsive')) {
+      const viewport = this.resources_.getViewport();
+      const viewportWidth = viewport.getWidth();
+      const viewportHeight = viewport.getHeight();
+      const responsiveHeight =
+        getResponsiveHeightForContext_(viewportWidth, viewportHeight);
+      const margins = this.margins_ || {};
```

Comment:
I think you need to deep copy this.margins_ or else the code below will be updating this.margins_. Suggest adding a function in https://github.com/ampproject/amphtml/blob/master/src/layout-rect.js called cloneMarginsChangeDef(marginsChange) that returns a deep copy of the argument. That way if someone ever changes LayoutMarginsChangeDef to include other objects they'll hopefully see that copy function and update it too.

Reply:
Done